game-core/src/main/java/games/strategy/triplea/ai/pro/util/MoveBatcher.java

```diff
@@ -0,0 +1,164 @@
+package games.strategy.triplea.ai.pro.util;
+
+import games.strategy.engine.data.Route;
+import games.strategy.engine.data.Unit;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * MoveBatcher implements an algorithm that batches AI moves together, reducing the number of
+ * distinct moves the AI does (and thus the visual delay) by combining moves together.
+ */
+public class MoveBatcher {
+  private static class Move {
+    private final ArrayList<Unit> units;
+    private final Route route;
+    private final HashMap<Unit, Unit> unitsToTransports;
+
+    private Move(
+        final ArrayList<Unit> units,
+        final Route route,
+        final HashMap<Unit, Unit> unitsToTransports) {
+      this.units = units;
+      this.route = route;
+      this.unitsToTransports = unitsToTransports;
+    }
+
+    Move(final ArrayList<Unit> units, final Route route) {
+      this(units, route, null);
+    }
+
+    Move(final Unit unit, final Route route, final Unit transportToLoad) {
+      this(
+          mutableSingletonList(unit),
+          route,
+          new HashMap<Unit, Unit>(Map.of(unit, transportToLoad)));
+    }
+
+    private static ArrayList<Unit> mutableSingletonList(final Unit unit) {
+      return new ArrayList<>(Collections.singletonList(unit));
+    }
+
+    boolean isTransportLoad() {
+      return this.unitsToTransports != null;
+    }
+
+    boolean canMergeWith(final Move other) {
+      return other.isTransportLoad() == isTransportLoad() && route.equals(other.route);
+    }
+
+    boolean mergeWith(final Move other) {
+      if (canMergeWith(other)) {
+        // Merge units and transports.
+        units.addAll(other.units);
+        if (isTransportLoad()) {
+          unitsToTransports.putAll(other.unitsToTransports);
+        }
+        return true;
+      }
+      return false;
+    }
+
+    void addTo(
+        final List<Collection<Unit>> moveUnits,
+        final List<Route> moveRoutes,
+        final List<Map<Unit, Unit>> unitsToTransports) {
+      moveUnits.add(units);
+      moveRoutes.add(route);
+      unitsToTransports.add(this.unitsToTransports);
+    }
+  }
+
+  private final ArrayList<ArrayList<Move>> moveSequences = new ArrayList<ArrayList<Move>>();
```

Comment:
Agree, for example the constructor ` Move(final ArrayList<Unit> units, final Route route) {`,
if we change the implementation type, then we'll have to refactor the constructor, perhaps unnecessarily. It also implies that we really want this exact type when perhaps an `Iterable` or `Collection` are the API.
There are other reasons to go with the interface type:
- it's a general expectation in Java to code to interfaces, "least surprise" principle
- rest of the code base is pretty consistent to use the most generic interface type possible

Reply:
So before I wrote a unit test, I actually had the API accept List instead of ArrayList. Then, in the unit test, I passed in List.of() which compiled just fine and gave runtime errors since the implementation relies on being able to add to it.

And also to iterate by index (see mergeSequences()) - so LinkedList would also not be a good choice.

Given that using ArrayList concrete types avoids both those issues (i.e. runtime errors if someone tries to pass in List.of() that's not modifiable and performance issues if someone tries to pass in LinkedList), I would we prefer we keep ArrayList as the type here.

Reply:
Interesting reasoning, though I don't know if I hold quite the same view. Let's say for example that someone else has a Set, or a different kind of list, to use the API presented they have to cast their object to an ArrayList, a very specific data type. That perhaps is not necessary as you don't actually need the API of an ArrayList, just the API of a List or a Collection, so you're forcing them to change their data type to suit your implementation needs.

The reveal that passing in an immutable list is interesting, I don't know if constraining the parameter is the right fix. We do have a convention to favor immutability and mark everything as final. If we use mutable data structures and then nothing is marked as final because it's all mutable, then the convention is without purpose. I think it's also then a bit striking as well that the input parameter is modified. 'Out' or 'in/out' parameters I'll suggest are best avoided, they're essentially side effects and force people using or reading usages of this code to know about the implementation. In part, we want anyone using this API, or reading code that uses it, to be unburdened with knowing the API. From a human scaling perspective, it's very limiting, to understand any code that uses this code, you really need to know "this" code, and so on, which means to modify anything at a top-level part of the abstraction, you have to be familiar with everything below it. It's a high-bar for entry, means we spend a lot of time examining underlying code before being able to make a modification.

I don't mean to be piling on, just conveying that these kind of decisions can add up to real maintenance cost in the form of having to "know" about a lot of code to be able to change just a few lines.

Another really easy way to fix the immutability issue is to avoid using an 'in' parameter, instead make of copy of the parameter, modify that, and then return the modified result.